CTC Decoding for JAX and Tensorflow

MaanasArora opened the pull request:
Hello,

This is WIP, but I wanted to create a PR for the changes so far as I continue to work. The PR currently contains:

- A Keras op for decoding using greedy and beam search strategies
- Implementation for both greedy and beam search strategies in TF
- Implementation for greedy strategy in JAX (beam search is WIP)

I will add torch support and unit testing. Any feedback is appreciated.

Thank you!

abhaskumarsinha commented:
Hello, can you add these codes in keras-nlp? I believe it'd be more helpful to have token search strategies in Keras NLP because all the LLMs to test these features are there only. 
Src: https://github.com/keras-team/keras-nlp 

@mattdangerw Please check!

fchollet commented on keras/backend/jax/nn.py:
```diff
@@ -622,3 +622,73 @@ def _iterate(prev, x):
+    merge_repeated=True,
+    mask_index=None,
+):
+    if strategy == "greedy":
```
This would need the fallback case `else:`

MaanasArora replied:
Yes, added, thanks.

fchollet commented on keras/backend/tensorflow/nn.py:
```diff
@@ -802,3 +802,62 @@ def ctc_loss(
+        )
+    else:
+        raise ValueError(
+            "Invalid strategy. Supported values are 'greedy' and 'beam_search'."
```
Print the strategy that was passed

MaanasArora replied:
Agreed and done, thanks!

fchollet commented on keras/backend/tensorflow/nn.py:
```diff
@@ -802,3 +802,62 @@ def ctc_loss(
+
+    sequence_length = tf.convert_to_tensor(sequence_length)
+    if strategy == "greedy":
+        decoded, probs = tf.nn.ctc_greedy_decoder(
```
You can just do `return tf.nn.ctc_greedy_decoder(...`

MaanasArora replied:
This is done, thanks for noticing.

fchollet commented on keras/backend/tensorflow/nn.py:
```diff
@@ -802,3 +802,62 @@ def ctc_loss(
+            blank_index=mask_index,
+        )
+    elif strategy == "beam_search":
+        decoded, probs = tf.nn.ctc_beam_search_decoder(
```
Same here

MaanasArora replied:
Done as well.

fchollet commented on keras/ops/nn.py:
```diff
@@ -1855,6 +1855,66 @@ def ctc_loss(target, output, target_length, output_length, mask_index=0):
+
+    if any_symbolic_tensors((inputs, sequence_lengths)):
+        raise NotImplementedError(
+            "CTC decoding is not supported in graph mode. "
```
This is a bit of a misleading message. You could simply "is not supported with KerasTensors. Use it inside the `call()` method of a Layer or the `predict_step` method of a model."

MaanasArora replied:
That's definitely a much better and more accurate message! Just replaced it. Thank you!

fchollet reviewed (commented):
Thank you for the PR!

> I believe it'd be more helpful to have token search strategies in Keras NLP because all the LLMs to test these features are there only.

This is an interesting suggestion -- the `Sampler` functionality in KerasNLP is similar. + @mattdangerw to provide a judgement call here.

mattdangerw commented:
I won't claim to be an expert on CTC classification (so may be missing lots here). But I would probably keep this separate from sampling strategies for open ended generation in KerasNLP. Seems like there's too many differences that might rear their head given that end tasks the two are solving are different.

E.g. I see `jax.lax.scan` here, but I don't think that would work for generation with early stopping on a "end token". For early stopping with a condition we need a `jax.lax.while_loop` (or a python loop with an inner compiled step).

So given that we are adding this here in part for backward compat with keras 2 (at least I think?), probably can continue to land as is. We can always stitch KerasNLP and these Keras utils together in an example.

fchollet commented:
> So given that we are adding this here in part for backward compat with keras 2 (at least I think?), probably can continue to land as is. We can always stitch KerasNLP and these Keras utils together in an example.

Sounds good -- let's keep them separate then.

@MaanasArora we'll be able to merge this soon! Please add unit tests.

MaanasArora commented:
Thank you @fchollet! I have added a unit test for correctness. I omitted the shape test to maintain consistency because `ctc_decode` does not support `KerasTensors`.

abhaskumarsinha commented:
I'm not an expert either tbh.

> E.g. I see jax.lax.scan here, but I don't think that would work for generation with early stopping on a "end token". For early stopping with a condition we need a jax.lax.while_loop (or a python loop with an inner compiled step).

Ah, I see. Thank you for pointing it.

fchollet commented on keras/ops/nn_test.py:
```diff
@@ -1870,6 +1870,40 @@ def test_ctc_loss(self):
+        score_labels = np.array([[-1.2], [-1.3], [-0.4]])
+
+        (decoded,), scores = knn.ctc_decode(
+            inputs, sequence_lengths=[3, 3, 1], strategy="greedy"
```
We should also check the different strategies (when available) and the behavior of the different arguments for the beam search strategy.

MaanasArora replied:
Just added a unit test for beam search when the backend is TF, thanks!

fchollet reviewed (commented):
Thanks for the update. Do you also intend to add the torch implementation? I think there's a built-in utility in torch that we could use. If not, you can add a `NotImplementError` in the torch backend.

MaanasArora commented:
Thank you for reviewing! Yes, I do intend to add the torch implementation. The built-in implementation seems to be in [torch audio](https://pytorch.org/audio/main/generated/torchaudio.models.decoder.CTCDecoder.html) rather than torch core, so I wondered if we should use it directly. If you would like, I can include both the JAX beam search implementation and the torch implementations in another PR.

fchollet commented:
Cool -- you can keep adding the functions to the current PR, that will make it easier to handle any potential API changes.

fchollet commented:
@MaanasArora are you still working on this PR?

MaanasArora commented:
For sure, thank you. I have pushed the changes I have so far. The tests for CTC beam search decode fail because of the way I am handling masked entries. In the way I am using `_pad` to replace `mask_index` on beam extensions, repetitions of the same character are not possible since there is no record at each time step of whether a path ends with `mask_index`. However, when we do use `mask_index` appropriately to extend sequences, at least once, we cannot use `jnp.unique` to accurately filter for unique paths, as paths with a final mask token and those without are treated differently.

I have been struggling to incorporate `jnp.unique` somehow because of simplicity, but given it seems difficult, I am now planning to use `jnp.lexsort` (or `lax.scan`, if needs be) to find unique paths instead, so we can use custom functionality to merge the scores for paths that are identical except for a final mask_index.

MaanasArora commented:
Hello, I was able to fix the issue I mentioned. I used a boolean array to decouple masking from the path arrays. I also addressed the effect of the sequence_length and merge_repetitions arguments. But there is still a bug regarding the storage of beam entries; the output differs from the TF implementation when the beam index is small. I will be working on that. Thank you!

MaanasArora commented:
Hello, I've fixed the bugs I mentioned, refactored for clarity, added JAX to the unit test for beam search, and also improved some of the unit testing. I'm planning to try to improve the code structure further if possible.

MaanasArora commented:
Just fixed another minor bug--it seems that `jnp.argsort` does not support the `order` parameter, without which sorting causes a discrepancy with the tensorflow implementation. I've flipped the inputs on the class axis and then flipped back to the right classes at the end as a workaround.

fchollet commented on keras/backend/jax/nn.py:
```diff
@@ -622,3 +622,226 @@ def _iterate(prev, x):
+    mask_index=None,
+):
+    inputs = jnp.array(inputs)
+    sequence_length = jnp.array(sequence_length)
```
Should we cast this to int32?

MaanasArora replied:
Oh yes, nice catch! Done.

fchollet commented on keras/backend/jax/nn.py:
```diff
@@ -622,3 +622,226 @@ def _iterate(prev, x):
+    mask_index=None,
+):
+    inputs = jnp.array(inputs)
+    sequence_length = jnp.array(sequence_length)
```
Same here?

MaanasArora replied:
Done!

fchollet commented on keras/ops/nn.py:
```diff
@@ -1855,6 +1855,67 @@ def ctc_loss(target, output, target_length, output_length, mask_index=0):
+        A tuple containing:
+
+        - a list of decoded sequences.
+
```
Remove line break

MaanasArora replied:
Done, thanks!

fchollet reviewed (commented):
Nice work!

sampathweb commented:
Please note that we refactored code to move all sources to `keras/src` and public API is generated in `keras/api` folder. Please rebase your code with master.

Please run `./shell/api_gen.sh` script when making changes that affect `keras_export`.

MaanasArora commented:
@sampathweb This is done; thank you for the information.

fchollet commented:
> +16,586 −6,035

It looks like you will need to rebase your PR on master to avoid extraneous changes.

MaanasArora commented:
Yes, this is done @fchollet

MaanasArora commented:
Hi @fchollet, so although `torchaudio` does have a CTC decoding implementation, it is not a dependency of Keras. Also, I am not sure if the numerics would be consistent with TF and (our replicated implementation in) JAX. Is it better to create a custom solution?

fchollet commented:
> Hi @fchollet, so although torchaudio does have a CTC decoding implementation, it is not a dependency of Keras. Also, I am not sure if the numerics would be consistent with TF and (our replicated implementation in) JAX. Is it better to create a custom solution?

It depends! 

1. If we want to use it, we can do an inline import of torchaudio in the backend function, except the ImportError, and display an error that asks users to install it. We this with torchvision in a few places.
2. We could also take a look at the implementation and judge how difficult it would be to simply fork it.
3. We could also simply convert the JAX implementation to torch. Would it be difficult?

I think a big factor here is whether we're getting the same numerics with the torchaudio version. If not, better to reimplement (based on the JAX code, and potentially using the torchaudio code as reference).

MaanasArora commented:
It seems torchaudio uses the Flashlight library under the hood. I don't seem to be able to get the same numerics, unfortunately. It should not be too difficult to convert the JAX code, though I'm not very familiar with torch's low-level functions yet.

fchollet commented:
> In case you'd like to release this first or open this up to others, we can probably create another PR for that? Thank you!

Sounds good -- please add the op function in `backend/torch/nn.py` and `backend/numpy/nn.py` and raise `NotImplementedError` in the functions. Also format the code via `sh shell/format.sh`. I'll do a full review tomorrow.

MaanasArora commented:
> Sounds good -- please add the op function in backend/torch/nn.py and backend/numpy/nn.py and raise 
> NotImplementedError in the functions. Also format the code via sh shell/format.sh. I'll do a full review tomorrow.

Both done! Thank you for reviewing!

fchollet reviewed (approved):
Looking great -- thank you for the contribution!

Will do a few minor docstring edits post-merge.
